Add `x` keyword, ``crop`` and ``_slice_me`` to `FunctionalDataAxis` and fixes

francisco-dlp opened the pull request:
### Description of the change

This PR is based on @jlaehne 's https://github.com/ericpre/hyperspy/pull/17 and is inspired by the original approach to solve the same problem in https://github.com/ericpre/hyperspy/pull/11. Originally I intended to fix a couple of things in https://github.com/ericpre/hyperspy/pull/17 , but then I realized that the original approach taken by @jlaehne  in  https://github.com/ericpre/hyperspy/pull/11 (that I argued against in https://github.com/ericpre/hyperspy/pull/11#issuecomment-568157095 and led @jlaehne to implement https://github.com/ericpre/hyperspy/pull/17) was better (my apologies @jlaehne).

With this PR a ``FunctionalDataAxis`` can be constructed:
* With an array, ``x``, and an expression
* With ``size`` and ``expression``, in which case ``x`` becomes ``np.arange(size)`` and is updated if ``size`` changes.

This also includes:

* Fixes for ``_slice_me`` and ``crop``
* A method to convert a linear axis to a FDA, see example below.
* The expression parameters are now traits
*  Initializing the expression parameters is now mandatory

This is WIP. I'll work on the documentation and tests once we decide if this is the right way to implement ``FunctionalDataAxis``.

### Progress of the PR
- [ ] Change implemented (can be split into several points),
- [ ] update docstring (if appropriate),
- [ ] update user guide (if appropriate),
- [ ] add tests,
- [ ] ready for review.

### Minimal example of the bug fix or the new feature
```python
>>> import hyperspy.api as hs
>>> s = hs.datasets.eelsdb(title="Magnesium")[0]
>>> s.axes_manager[0].convert_to_functional_data_axis(expression="1.2398E3 / x", units="nm", name="wavelength")
>>> s.isig[::-1].plot()
```
![Figure_Magnesium_carbonate_Signal](https://user-images.githubusercontent.com/989719/71996059-da96da00-323b-11ea-96a7-4866ad97dc85.png)

francisco-dlp commented:
I have changed the type of the ``x`` parameter of ``FunctionalDataAxis`` from array to ``BaseDataAxis``. This has the advantage of using the same mechanism as ``LinearDataAxis`` for cropping and slicing when ``x`` is a ``LinearDataAxis``. Another advantage is that we could now use ``x`` to plot a secondary axes using the [new feature](https://matplotlib.org/gallery/subplots_axes_and_figures/secondary_axis.html) in matplotlib.

jlaehne commented on hyperspy/axes.py:
```diff
@@ -688,66 +682,57 @@ def crop(self, start=None, end=None):
+        if x is None:
+            if size is t.Undefined:
+                raise ValueError("Please provide either `x` or `size`.")
+            self.x = LinearDataAxis(scale=1, offset=0, size=size)
```
Defaulting `offset=1` would avoid `divide by zero` error when using a reciprocal function.

jlaehne reviewed (commented):
OK, seems to make more sense now. Slicing starts to work also for FDA, but as commented an `isig` operation modifies `x` permanently.

francisco-dlp commented:
Good catch @jlaehne, I'll fix it as soon as I have time this week.

Regarding ``offset=1``, I agree that that would prevent getting an exception, but exceptions are sometimes good because it get users thinking on what they are doing. After thinking on this from the users point of view, I think that the best approach is not to have a default axis. In this way users are obliged to know what they're doing and they won't get an exception unless they get it wrong. What do you think?

jlaehne commented:
Good point that we do not need to pamper the user too much.

Bye, the way - nice solution with `x = BaseDataAxis` and the potential for dual axis plotting.

jlaehne commented:
@francisco-dlp I lsot a bit track, did you fix the issue with `isig`?

ericpre committed:
Clarify output load functions in docstring.

gertjan-devos committed:
Initial working Phenom elid reader implementation

pquinn-dls committed:
Merge branch 'RELEASE_next_minor' of https://github.com/hyperspy/hyperspy.git into xrf_components

pquinn-dls committed:
updated docstrings to improve formatting

pquinn-dls committed:
Fixed docstrings and centered the Doniach lineshape

pquinn-dls committed:
Updated with changes from Francisco

francisco-dlp committed:
Merge pull request #2326 from pquinn-dls/xrf_components

asymmetric lineshapes for eds and xrf fitting

gertjan-devos committed:
Implement ELID test cases

tjof2 committed:
Improve s.decomposition()

Fix garbled GPL header and format

Rename to features/samples from variables/trials

Update tests to check shapes

Test store_r, and speed up tests by shrinking matrix

Reorder comparisons to avoid 'FutureWarning: elementwise comparison failed'

Rename I -> K to avoid ambiguity over naming

Re-order imports, tidy-up s.decomposition() docstring

Update svd_pca to take new svd_solver arguments as per sklearn

Update RPCA and tests to use new svd_solver

Fix for output_dimension=None and warning message

Re-order and simplify s.decomposition()

Use np.errstate instead of np.seterr

Update elbow function to avoid raising divide-by-zero warnings, and vectorize rather than loop

Rename private functions and improve docstring

Replace logger warnings with INFO level for parameter defaults

Update tests to cover warnings, and temporarily disable MLPCA waiting on PR #2352

Add initial code for using sklearn objects

Check non-negative first and raise error, plus additional docstring fixes

Add more general centering & whitening method for PCA and ZCA whitening

Add return_info and print_info to bss()

Update MLPCA to use fast SVD solver if available

Update tests to remove warnings

Update orthomax location

Optimize orthomax approach

Tidy up sklearn imports

Fix orthomax error causing failing tests

Add orthomax BSS test

Update test and check iterations > 0

Better coverage of SVD PCA

Mark whole test classes since it all relies on sklearn

Support sklearn pipelines and gridsearchCV

Fix RPCA to allow us to project low-rank only

Check algorithm unknown raises error

Move tests/mva/* to tests/learn/* for consistency

Tests to check RPCA and RNMF works with signals not just arrays as per #1557 and #1558

Replace store_r by store_error for clarity and consistency

Move whitening to its own file

Remove polyfit

Update warnings

Test reprojection warnings

gertjan-devos committed:
Add user guide documentation the .elid format

tjof2 committed:
Initial commit of re-written function

tjof2 committed:
Pass kwargs to mlpca

Check fast_mlpca too

tjof2 committed:
Fix docstring

tjof2 committed:
Add unit test for MLPCA

tjof2 committed:
Correct typo in args, and also allocate memory first

ericpre committed:
Fix error message.

ericpre committed:
Use f-string in error message and fix missing space.

jasoncmyers committed:
Change exponential component to use geometric mean instead of two endpoints

ericpre committed:
Fix typo.

jasoncmyers committed:
Add exponential unit tests and clean up (PEP8)

jasoncmyers committed:
Merge branch 'RELEASE_next_minor' of https://github.com/hyperspy/hyperspy into DEV_next_minor

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

tjof2 committed:
Replace scipy.signal.medfilt with faster scipy.ndimage.median_filter

Calculate optimal FFT size better

tjof2 committed:
add clarifying comment

jasoncmyers committed:
Add exponential to remove_background docstring and tests; add RuntimeWarning to caught exceptions for Exponential.estimate_parameters()

jasoncmyers committed:
Merge branch 'RELEASE_next_minor' of https://github.com/hyperspy/hyperspy into DEV_next_minor

tjof2 committed:
Use rfft when not doing sub-pixel alignment, and fix import error

ericpre committed:
Run a garbage collection cycle when tearing down the test_remove_background.py module. The test suite was hanging on appveyor python 3.8 when running samfire test.

tjof2 committed:
Add wrapper around scipy next_fast_len to handle different versions

tjof2 committed:
Use wrapped function

tjof2 committed:
Update optimal FFT size to not use powers of two

tjof2 committed:
Missed FFT size

tjof2 committed:
Simplify sub-pixel test

tjof2 committed:
Option to use rfft

tjof2 committed:
Revert "Option to use rfft"

This reverts commit 085438bd931310d0bdb94a012fcb8e58ac99344a.

tjof2 committed:
option to return real

tjof2 committed:
Revert "Revert "Option to use rfft""

This reverts commit 06946c31f359e66a2407fe05f7830b478710b043.

tjof2 committed:
Rename arg

tjof2 committed:
add garbage collection

tjof2 committed:
Revert "add garbage collection"

This reverts commit 9421f10bb80f744f6a51fe731d0bd31699120b95.

tjof2 committed:
Run a garbage collection cycle when tearing down the test_remove_background.py module. The test suite was hanging on appveyor python 3.8 when running samfire test.

tjof2 committed:
Revert this 2*

tjof2 committed:
Reduce size of samfire tests for memory usage (still passes)

tjof2 committed:
Update test_samfire to reduce memory consumption, hanging on appveyor

tjof2 committed:
Reduce num_workers to 1 in samfire

tjof2 committed:
Reduce num_workers to 1 in samfire

tjof2 committed:
add back in xfail with better explanation

tjof2 committed:
add back in xfail with better explanation

tjof2 committed:
add in gc.collect()

tjof2 committed:
add in gc.collect()

tjof2 committed:
Close down parallel pools from samfire

tjof2 committed:
Close down parallel pools from samfire

tjof2 committed:
Remove unnecessary check

tjof2 committed:
Remove unnecessary check

tjof2 committed:
Change dtype to float32 for memory

tjof2 committed:
Change dtype to float32 for memory

tjof2 committed:
Update test dtypes

samfire tests work with float16 dtype also

tjof2 committed:
Update test dtypes

samfire tests work with float16 dtype also

tjof2 committed:
Put back in change as per @ericpre

francisco-dlp committed:
Merge pull request #2354 from ericpre/fix_typo

Fix typo docstring.

ericpre committed:
Merge pull request #2358 from tjof2/improve-align-2d

Performance improvements for estimate_shift_2D() and other FFTs

tjof2 committed:
Merge branch 'RELEASE_next_minor' into fix-mlpca-method

ericpre committed:
Remove appveyor and add azure pipelines.

ericpre committed:
Update travis.

ericpre committed:
Fix building doc and remove image comparison on travis.

ericpre committed:
Fix typo in azure pipeline.

ericpre committed:
Tweak travis.

ericpre committed:
Use freetype 2.10 to make image comparison fail.

ericpre committed:
Simplify azure pipeline script.

ericpre committed:
Revert to freetype 2.9 to make the test suite pass now that we are sure the results of image comparison are uploaded as artefact upon failure.

ericpre committed:
Add missing optional dependencies in conda environment.

ericpre committed:
Remove building wheel on travis, this is now done on azure pipeline (publish artefact on azure pipeline and push to github release).

francisco-dlp committed:
Fix wrong error when zlp not None and threshold None

francisco-dlp committed:
Add mean_free_path keyword

francisco-dlp committed:
Add angular corrections and MFP estimation

francisco-dlp committed:
Rename functions

tjof2 committed:
Fix log(0) errors in estimate_elbow_position

ericpre committed:
Merge pull request #2365 from tjof2/fix-elbow-method

Fix log(0) errors in estimate_elbow_position

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into fix-mlpca-method

tjof2 committed:
Add MLPCA/fast MLPCA to tests

tjof2 committed:
avoid np.diag()

ericpre committed:
Update the CI scripts repo to hyperspy/ci-scripts.

francisco-dlp committed:
Add R. Egerton relativistic correction

francisco-dlp committed:
Cite only Egerton since we use the modified equations

ericpre committed:
Update service connection name to pull azure templates.

francisco-dlp committed:
Merge pull request #2363 from ericpre/add_azure_pipelines

Add azure pipelines and remove appveyor

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into ENH_eels_thickness

ericpre committed:
Update plot testing in the dev guide.

ericpre committed:
Skip test which failed quite often on python 3.6.

francisco-dlp committed:
Add tests

francisco-dlp committed:
Add data files

francisco-dlp committed:
Add UG doc

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into ENH_eels_thickness

ericpre committed:
Skip mrcz test on windows and complex type.

jasoncmyers committed:
Merge branch 'RELEASE_next_minor' of https://github.com/hyperspy/hyperspy into DEV_next_minor

ericpre committed:
Install all optional dependencies when building doc to avoid skipping docstring in optional modules.

jasoncmyers committed:
Update Exponential model exceptions to handle signal range of only a single point

ericpre committed:
Merge pull request #2355 from jasoncmyers/DEV_next_minor

Add exponential component to the options for remove_background()

ericpre committed:
Mark mrcz test xfail for dtype on windows (not only complex dtype).

francisco-dlp committed:
Add quantity info to metadata

francisco-dlp committed:
Add tests

thomasaarholt committed:
updated docs on smart_fit

docstring

francisco-dlp committed:
Improve UG docs, add Iakoubovskii2008

francisco-dlp committed:
Add warning when not applying angular corrections

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into fix-mlpca-method

ericpre committed:
Use skipif instead of xfail and fix typo.

ericpre committed:
Add documentation about continuous intergration in the developer guide.

tjof2 committed:
add azure, remove appveyor & saythanks

tjof2 committed:
typo

ericpre committed:
Try to increase the time in the sleep when testing mrcz async, remove skip test on windows.

francisco-dlp committed:
Merge pull request #2371 from tjof2/update-badge

Add Azure Pipelines badge to README.rst

francisco-dlp committed:
Fix travis badge

Thanks @tjof2

thomasaarholt committed:
renamed variables

francisco-dlp committed:
Use Poisson noise in test

tjof2 committed:
Merge pull request #1 from francisco-dlp/fix-mlpca-method

Use Poisson noise in test

francisco-dlp committed:
Merge pull request #2376 from thomasaarholt/eels_A_r

Renamed EELSCLEdge private attributes

francisco-dlp committed:
Merge pull request #2370 from thomasaarholt/smart_fit

Updated docs on smart_fit

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into fix-mlpca-method

tjof2 committed:
Remove .dot() and ensure result is low-rank

francisco-dlp committed:
Merge pull request #2352 from tjof2/fix-mlpca-method

Fix maximum-likelihood PCA method

tjof2 committed:
Further improvements to mva.py

Normalize option not supported

Fix indents

Check return_info and print_info for BSS

Remove GoDec G matrix for memory

Consistency between ORNMF and ORPCA for batch size etc.

Replace np.dot with @

Fix init method

Check for sklearn convergence warnings

Consistent ORNMF and ORPCA tests

Increase test coverage and add a __repr__ method to LearningResults()

Improve docstrings

Update tests

ericpre committed:
Correct typo dev guide, thanks @tjof2!

francisco-dlp committed:
Merge pull request #2369 from ericpre/doc_test_suite

Doc test suite

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

# Conflicts:
#	doc/dev_guide/testing.rst

gertjan-devos committed:
Resolved merge conflict

francisco-dlp committed:
Re-enable codecov

francisco-dlp committed:
ericpre -> hyperspy

francisco-dlp committed:
Add missing newline

francisco-dlp committed:
Add missing newline

gertjan-devos committed:
Processed review comments

francisco-dlp committed:
Fix?

ericpre committed:
Merge pull request #2351 from djeedjay/phenom_elid_reader

Phenom elid reader

francisco-dlp committed:
Print PYTEST_ARGS variable for debugging

francisco-dlp committed:
Fix?

francisco-dlp committed:
FIX?

francisco-dlp committed:
FIX?

francisco-dlp committed:
FIX?

francisco-dlp committed:
Fix?

francisco-dlp committed:
Fix?

francisco-dlp committed:
Fix?

francisco-dlp committed:
Previous attempt worked, restore pip selector

francisco-dlp committed:
Only run coverage when computing it

francisco-dlp committed:
Fix missing "

francisco-dlp committed:
Remove unnecessary export

francisco-dlp committed:
Turn sphinx warnings into errors

In this way the CI fails.
See https://github.com/jdillard/continuous-sphinx/blob/master/.travis.yml


francisco-dlp committed:
Fix sphinx-build

francisco-dlp committed:
Revert "Fix sphinx-build"

This reverts commit 0958eab300acfe1d954fa738635557924295fe68.

francisco-dlp committed:
Revert "Turn sphinx warnings into errors"

This reverts commit 4079cc7cfe9eed0bc1cee0342498e06aa157df37.

tjof2 committed:
Improve decomposition documentation as part of #1159

Undo accidental change

Add image

Update docstring references

Add tests for lazy decomposition

ericpre committed:
Merge pull request #2378 from francisco-dlp/ENH_reenable_codecov

(Re)enable code coverage and fail on sphinx warnings

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into clarify_user_guide_load

# Conflicts:
#	hyperspy/_signals/signal1d.py

ericpre committed:
Tweak documentations.

francisco-dlp committed:
Merge pull request #2379 from ericpre/fix_error_message_toolkit

Fix error message toolkit

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into ENH_eels_thickness

tjof2 committed:
Move pytest.ini to setup.cfg

tjof2 committed:
Add flake8 linting options

tjof2 committed:
Move coveragerc to setup.cfg

tjof2 committed:
add blank line

tjof2 committed:
Update opts to print all

tjof2 committed:
Update docs

tjof2 committed:
Explicitly search only test path

tjof2 committed:
Bring lazy decomposition inline with normal decomposition

Correctly support FastICA objects

Deprecate variables/trials in favour of clearer notation

Update documentation

Bugfix

tjof2 committed:
Register markers differently

ericpre committed:
Merge pull request #2381 from tjof2/setup-cfg

Combine pytest.ini and .coveragerc into setup.cfg

tjof2 committed:
Missed filename

tjof2 committed:
Bugfix with import

tjof2 committed:
Rename back to old directory for cleaner diff

tjof2 committed:
Make svd_flip optional to preserve plotting tests

tjof2 committed:
Fix arguments for svd_flip

tjof2 committed:
Rename function to avoid collision with kw argument

tjof2 committed:
Update test images due to svd_flip

tjof2 committed:
Fix sklearn error

tjof2 committed:
Catch linalg error when inverting

tjof2 committed:
add sklearn skipif

tjof2 committed:
Missed colon

tjof2 committed:
add additional sklearn skipif

tjof2 committed:
additional skipif for sklearn

tjof2 committed:
Test cumulative_explained_variance_ratio plot too

tjof2 committed:
Replace .format() with f-strings as per @ericpre

tjof2 committed:
Fix documentation as per @ericpre

tjof2 committed:
Update docs

tjof2 committed:
Fully document TEM .decomposition()

tjof2 committed:
Fix doc warning

tjof2 committed:
First comments resolved from @francisco-dlp

tjof2 committed:
Lower->Upper for whitening

tjof2 committed:
Revert "Lower->Upper for whitening"

This reverts commit 1b412e0d9289adcbee22d1de903351523f6bd5dd.

tjof2 committed:
Note on scree plots

ptim0626 committed:
Add a method to get EELS edges from a given energy

This method returns EELS edges from a given energy window and it is in a list
arranged in the order of the energy difference to the specified energy.

tjof2 committed:
Update docs

ptim0626 committed:
Prevent using a negative energy width

Usually this is by accident, so raising error will avoid confusion.

francisco-dlp committed:
Merge pull request #2383 from tjof2/improve-decomposition-learn

Improve machine learning API and documentation

ptim0626 committed:
Write tests for the method get_edges_near_energy

ptim0626 committed:
Add the user guide for get_edges_near_energy()

tjof2 committed:
Move test_mpfit to tests/ (missed in #2381)

tjof2 committed:
Tidy up README.rst

tjof2 committed:
Undo link change

francisco-dlp committed:
Merge pull request #2388 from tjof2/tidy-readme-fix-test

Tidy Readme & fix test path

ptim0626 committed:
Change the default energy width to 10 from 1.

ptim0626 committed:
Clarify the meaning of arguments in the user guide

ericpre committed:
Merge pull request #2387 from ptim0626/eels_edge_near_energy

Get EELS edges from a given energy window

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

# Conflicts:
#	README.rst
#	doc/dev_guide/coding_style.rst
#	doc/dev_guide/testing.rst
#	doc/dev_guide/writing_docs.rst

francisco-dlp committed:
Remove Ray's gamma correction

francisco-dlp committed:
Add iMFP_TPP2M

francisco-dlp committed:
Rename beam_energy -> electron_energy

francisco-dlp committed:
Add copyright

francisco-dlp committed:
Fix citations, add Eq number, thanks @ericpre

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into ENH_eels_thickness

francisco-dlp committed:
Add returns to docstrings

ptim0626 committed:
Update user guide to reflect the relocation

tjof2 committed:
Update map_iterate and align2D

Add unused kwarg to lazy _map_iterate

Update map and all uses

Update docstrings

Update user guide

tjof2 committed:
Add missing tests

ericpre committed:
Fix import error scikit-image 0.17.0.

francisco-dlp committed:
Merge pull request #2393 from ericpre/fix_skimage0.17.0

Fix import error with scikit-image 0.17.0

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into RELEASE_next_minor

ericpre committed:
Merge pull request #2368 from francisco-dlp/ENH_eels_thickness

Absolute thickness estimation by EELS log-ratio method

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into improve-parallel-threads

tjof2 committed:
Need 'if ragged is False' rather than 'if not ragged' since ragged can also be None

ptim0626 committed:
Resolve merge conflict in test_eels

ptim0626 committed:
Remove leftover markers from merge conflict

ptim0626 committed:
Add get_edges_near_energy as a method in EELS signal

ptim0626 committed:
Update user guide for the usage from EELS signal

francisco-dlp committed:
Merge pull request #2391 from tjof2/improve-parallel-threads

Improve thread-based parallelism behaviour

ptim0626 committed:
Reverted 76595dc and 4cfe225

francisco-dlp committed:
Merge pull request #2390 from ptim0626/relocate_get_near_edge_energy

Relocate get_edges_near_energy()

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into non-linear-axis_FunctionalDA

francisco-dlp committed:
Fix test tolerance

jlaehne committed:
corrected default in plot_spectra docstring

jlaehne committed:
Typos, style in docstrings and messages.

jlaehne committed:
Docstring style according to numpy.

francisco-dlp committed:
Avoid test failure due to ipywidgets

francisco-dlp committed:
Merge pull request #2397 from jlaehne/plot-utils-docstrings

Plot utils docstrings

francisco-dlp committed:
Fix BSS test

tjof2 commented on hyperspy/exceptions.py:
```diff
@@ -190,6 +190,15 @@ def __init__(self, navigation_size, expected_navigation_size):
+class NonLinearAxisError(Exception):
+
+    def __init__(self):
+        self.error = ("Non linear axis are not supported.")
```
Should be either "axis is" or "axes are"

jlaehne replied:
OK

jlaehne commented:
There is still a problem with `isig`. For this minimal example:
```
S = hs.signals.Signal1D(data=np.arange(10))
S.axes_manager[0].convert_to_functional_data_axis(expression = '1 / (x+1)')
S.isig[0.6:0.1].plot()
S.isig[0.6:0.1].plot()
```
Plot works the first time, but not the second time.
`S.isig[0.6:0.1].data.size` returns `8` no matter how often you call it.
However, `S.isig[0.6:0.1].axes_manager[0].axis.size` will decrease by `1` for every time you call it (It always looses the first value of the axis vector!).

francisco-dlp committed:
Fix axis size bug, thanks @jlahene

francisco-dlp committed:
Fix DataAxis.size not updated

francisco-dlp committed:
Use non-uniform instead of non-linear axes

francisco-dlp commented:
Thanks @jlaehne, I think that it should be fixed now.

Notice also that I've just changed every mention to *non-linear* axes to non-uniform axes, that is the most widespread denomination.

francisco-dlp committed:
linear -> uniform

francisco-dlp committed:
Make create_axis private

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into clarify_user_guide_load

# Conflicts:
#	doc/user_guide/mva.rst
#	hyperspy/learn/mva.py

ericpre commented:
Should this PR go to the `non_uniform_axes` branch?

ericpre commented on doc/user_guide/axes.rst:
```diff
@@ -0,0 +1,484 @@
+The navigation and signal dimensions
+------------------------------------
+
+In HyperSpy, the data is interpreted as a signal array and, therefore, the data
+axes are not equivalent. HyperSpy distinguishes between *signal* and
```
I would remove the first sentence because it is not clear at all.

jlaehne replied:
OK

ericpre commented on doc/user_guide/axes.rst:
```diff
@@ -0,0 +1,484 @@
+In HyperSpy, the data is interpreted as a signal array and, therefore, the data
+axes are not equivalent. HyperSpy distinguishes between *signal* and
+*navigation* axes and most functions operate on the *signal* axes and
+iterate over the *navigation* axes. For example, an EELS spectrum image (i.e.
```
Make the EELS spectrum image example more generic.

jlaehne replied:
I think a specific example is helpful, added a list of other examples and also the 2D case.

ericpre commented on doc/user_guide/axes.rst:
```diff
@@ -0,0 +1,484 @@
+    <Signal2D, title: , dimensions: (30|20, 10)>
+
+
+Axes storage and ordering
```
I would move this section to the end, because this is a written for a more advanced usage of the `DatAxis`.

jlaehne replied:
OK

ericpre commented on doc/user_guide/axes.rst:
```diff
@@ -0,0 +1,484 @@
+When data is acquired sequentially, it is usually stored in acquisition order.
+When a dataset is loaded, HyperSpy generally stores it in memory in the same
+order, which is good for the computer. However, HyperSpy will reorder and
+classify the axes to make it easier for humans. Let's imagine a single numpy
```
I am not sure this is correct for all IO plugins? Mention the `optimize` argument of some functions?

ericpre commented on doc/user_guide/axes.rst:
```diff
@@ -0,0 +1,484 @@
+* ``low_index`` (int) and ``high_index`` (int) are the first and last index.
+* ``low_value`` (int) and ``high_value`` (int) are the smallest and largest
+  value.
+* The ``axis`` (array) vector stores the values of the axis points. However,
```
I think using the term `vector` will be confusing for many users, because they will think of it as an spatial/euclidean vector and not as a vector in the numpy sense. If we keep it, it needs to be explained.
In my opinion, we should mentioned it when introducing what the `axis` attribute is but this should be secondary, not be a defining element of `axis` and not necessary to understand `vector` in this context to understand the rest.

jlaehne replied:
using array instead of vector

ericpre commented on hyperspy/drawing/mpl_he.py:
```diff
@@ -99,7 +99,7 @@ def plot_navigator(self,
             sl.data_function = self.navigator_data_function
             sl.set_line_properties(color='blue',
-                                   type='step')
+                                   type='step' if axis.is_uniform else 'line')
```
Why changing 'step' to `line`?

ericpre commented on hyperspy/signal.py:
```diff
@@ -2086,9 +2087,17 @@ def get_dynamic_explorer_wrapper(*args, **kwargs):
         if not isinstance(navigator, BaseSignal) and navigator == "auto":
-            if (self.axes_manager.navigation_dimension == 1 and
+            if (self.axes_manager.navigation_dimension > 1 and
+                    np.any(np.array([not axis.is_uniform for axis in
```
This should be documented.

ericpre commented on hyperspy/signal.py:
```diff
@@ -3552,14 +3628,16 @@ def ifft(self, shift=None, return_real=True, **kwargs):
         ----
-        For further information see the documentation of
-        :py:func:`numpy.fft.ifftn`
+        Requires a linear axis. For further information see the documentation
```
linear -> uniform

jlaehne replied:
OK

ericpre commented on hyperspy/signal.py:
```diff
@@ -3937,25 +4015,35 @@ def map(
 
+        Note
+        ----
+        Currently requires a linear axis.
```
linear -> uniform

jlaehne replied:
OK

ericpre commented on hyperspy/signal.py:
```diff
@@ -3937,25 +4015,35 @@ def map(
+        else:
+            # Check if the signal axes have inhomogeneous scales and/or units and
+            # display in warning if yes.
+            scale = set()
```
A bit of matter of taste, but I usually find list comprehension more readable:
```python
scale = set([axis.scale for axis in self.axes_manager.signal_axes])
```

ericpre commented on hyperspy/axes.py:
```diff
@@ -1343,24 +1743,33 @@ def _get_dimension_str(self):
                 self._get_dimension_str())
-        ax_signature = "% 16s | %6g | %6s | %7.2g | %7.2g | %6s "
+        ax_signature_uniform = "% 16s | %6g | %6s | %7.2g | %7.2g | %6s "
+        ax_signature_non_uniform = "% 16s | %6g | %6s |  non-uniform axis  | %6s "
```
There is one space too much with `non-uniform axis`: the columns are not aligned between uniform and non-uniform axis.

jlaehne replied:
OK

ericpre reviewed (commented):
I have lost track of this but i have tried to catch up and comments on this PR. Most of my comment are most likely not related to this PR, in which case, we come back to them in another PR.

I can do another review once this is ready.

francisco-dlp committed:
Merge pull request #2350 from ericpre/clarify_user_guide_load

Clarify documentation load functions

francisco-dlp committed:
Replace NonUniformAxisError with NotImplementedError

jlaehne commented:
> Thanks @jlaehne, I think that it should be fixed now.
> 
> Notice also that I've just changed every mention to _non-linear_ axes to non-uniform axes, that is the most widespread denomination.

Looks good now.

jlaehne committed:
more changes to 'uniform' + docs

jlaehne commented:
I opened a PR to this branch that addresses part of the issues raised by @ericpre and the typo found by @tjof2. Also it corrects more mentions of the 'linear axis' to 'uniform axis'

francisco-dlp committed:
Raise NotImplementedError for most EELS and holography methods

francisco-dlp committed:
Merge pull request #45 from jlaehne/pr2299-doc-corrections

Corrections to 'uniform' + documentation updates

francisco-dlp committed:
Merge branch 'non-linear-axis_FunctionalDA' of https://github.com/francisco-dlp/hyperspy into non-linear-axis_FunctionalDA

francisco-dlp committed:
Adapt update_axes_attributes_from

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into non-linear-axis_FunctionalDA

francisco-dlp committed:
CircleROI not implemented

jlaehne commented:
Looking again at the terminology, I find that both "non uniform" and "non linear" seem to be used. Google even has more hits for the latter, but I agree that "non uniform" sounds more intuitive, because we are basically talking about non uniform data spacing along the axis.

jlaehne commented:
Working on examples for the documentation, I realized that only the following works, which is a bit weird as you have to define an axis in the dictionary:
```
dict0 = {'expression': 'a / x + b', 'a': 100, 'b': 10, 'x': UniformDataAxis(size=10,offset=10,scale=0.1)}
s = hs.signals.Signal1D(np.ones(500), axes=[dict0])
```

Might make sense to add the functionality to define a nested dictionary (which might actually be needed for IO functionality?) so that the following would also work:
```
dict0 = {'expression': 'a / x + b', 'a': 100, 'b': 10, 'x': {'size': 10, 'offset': 10, 'scale': 0.1}}
s = hs.signals.Signal1D(np.ones(500), axes=[dict0])
```
Similarly, `get_axis_dictionary` currently returns: `... 'x': <Unnamed axis, size: 500>` and it might make sense to have it output the dictionary of that axis? (which you of course can already get if you specifically ask for the dictionary of the `x`-attribute: `s.axes_manager[0].x.get_axis_dictionary()`)

In a way that is linked to what #2345 is about, to be able to also initialize the signal using: 
```
axis = FunctionalDataAxis[expression='a / x + b', a=100, b=10, x=UniformDataAxis(size=10,offset=10,scale=0.1))
s = hs.signals.Signal1D(np.ones(500), axes=[axis])
```

francisco-dlp commented:
Regarding non-linear vs non-uniform, you are right @jlaehne. The nomenclature issue is that the axes are not linear or uniform. We probably should speak instead of *non-linear axis scale* and *non-uniform axis spacing*. I was inclined to use non-uniform because, in the multi-dimensional case, people speak about functions that operate on non-uniform grids e.g. [NFFT](https://en.wikipedia.org/wiki/Non-uniform_discrete_Fourier_transform). So, I was concerned that, by not using non-uniform, people where going to miss that part of the literature. However, now I don't know which one is better.

Concerning how to initialize `FunctionalDataAxis`, I think that you are right too, but I have to think a bit more about it.

jlaehne commented:
Well you can consider a logarithmic scale to be non-linear as well, even if your data is on a linear grid. Thinking about it, I am happy with non-uniform. It was just that the concept was running as non-linear for so long. Any oppinion from @ericpre?

I realized that we currently mix 'non-uniform' and 'non uniform'. We should probably stick to the former. Should I try to make that consistent across the files?

francisco-dlp commented:
I am starting to think that this is not at all ready to release even if we mention in the documentation that these features are in beta-state. Not only there are still a large number of functions that don't work on non-uniform axes, but also I think that the axes themselves need some more thought and work. My main current concern is that the current implementation assumes that the data represents variables sampled at discrete points. However, when `metadata.Signal.binned=True`, that no longer holds. For data on a linear scale is doesn't matter because we can always divide by the scale, but for non-linear scales we need to store more information in the axes. In particular, we must store the left and right position of the bins or one of those and their size. The fact that this is not yet implemented is very concerning because we know that, if we implement saving of signals with non-uniform axes now, we'll need to change the format again as soon as we implement the axes properly.

So, what about going back to the original plan of not including this in v1.6 and making a beta release of v1.7 as soon as we implement properly the feature for binned data? Both conda-forge and pypi support beta releases, so it shouldn't be a problem for external packages that require non-uniform axes to depend on a beta release.

francisco-dlp committed:
Some more NotImplementedErrors

ericpre commented:
> Well you can consider a logarithmic scale to be non-linear as well, even if your data is on a linear grid. Thinking about it, I am happy with non-uniform. It was just that the concept was running as non-linear for so long. Any oppinion from @ericpre?

I don't have any strong opinion on this, I would tend to agree that "non-uniform" may be better.

> 
> I realized that we currently mix 'non-uniform' and 'non uniform'. We should probably stick to the former. Should I try to make that consistent across the files?

Yes, indeed "non-unifrom axis" is more correct than "non uniform axis" and making the documentation more consistent is never a bad thing! :)


> So, what about going back to the original plan of not including this in v1.6 and making a beta release of v1.7 as soon as we implement properly the feature for binned data? Both conda-forge and pypi support beta releases, so it shouldn't be a problem for external packages that require non-uniform axes to depend on a beta release.

I agree on the binning issue and on the fact we should avoid making a format version and implement saving file from the beginning.

Even if the the non linear axis is not fully working at the moment, it doesn't break anything else as far as we know, so I am not convinced we need to do a separate beta release? Doing a normal single release would be more manageable and save us time but also user can use it more readily. The only reason we may need to do a beta release may be for 2.0 because they will be a lot of API changes and we may breaks a few things... so user will need to opt in to use the beta version and they will know they may be taking risk to have things broken!

jlaehne commented:
> Yes, indeed "non-uniform axis" is more correct than "non uniform axis" and making the documentation more consistent is never a bad thing! :)

Well, it were just a few instances in the end, I included the changes in a PR to this branch working on the FDA documentation.

> I agree on the binning issue and on the fact we should avoid making a format version and implement saving file from the beginning.

Ah, I did not have binned signals in mind as I am currently not using them. For the rest, I really had the impression that we are closing in and could use the momentum of the upcoming release to get it done. I have a few colleagues hesitant to adopt hyperspy as long as this functionality is not available and it pops up in the issues every now and then as well.

Do we really need to permanently store the bins in the axis or would it be enough to calculate the bins when needed by providing functions (e.g. `get_bins`) for `DataAxis` and `FunctionalDataAxis` classes, respectively, that calculate the bins from the axis (start value and size) when needed? The latter could simply be added later without changing the API or IO functions, while for the moment we raise exceptions when operating on binned signals. The advantage would also be that we do not have to add so much code to axis.py and dinstinguish in how binned and unbinned signals are saved. The metadata flag would remain the only difference. The functions where it is relevant would have to dinstinguish the two types, but they have to do so anyway.

francisco-dlp commented:
@ericpre, I agree with you, a beta release will be so rarely used that it is probably not worth the effort. Also, it is not reasonable to wait until all  the features work properly with non-uniform axes as this will take a long time. Also, by including it in a standard release we will boost the development of the non-uniform axes features, since users will be motivated to contribute their fixes once they discover that they are 1 feature away of getting their analysis done ;) On the other hand, in its current state, I think that too many things are broken and some important parts are half-cooked.

My main concern are "binned" signals, which are incidentally the most common kind of signals among HyperSpy users. Currently even `plot` is broken for them, since in their case it is misleading to plot e.g. counts vs energy when the bin size changes. We should plot counts / bin_size vs energy instead. This means that people cannot even use HyperSpy to load a dataset a have a look at it in a reliable way. Image plots suffer from the extra issue that currently HyperSpy plots them on an uniform grid. I think that we should usematplotlib's [NonUniformImage](https://matplotlib.org/3.2.1/gallery/images_contours_and_fields/image_nonuniform.html) or the better documented `pcolormesh` instead.

Hence, I will be comfortable releasing this when:

- [ ] Plot works properly for binned signals and images
- [ ] Binned versions of the axes are implemented
- [ ] Saving the axes in the hspy format is implemented
- [ ] Curve fitting on binned signals works (currently it doesn't)
- [ ] Most things that don't work on non-uniform axes are identified and an informative `NotImplementedError` is raised.

Low hanging remaining issues that we could consider fixing before releasing: 

- [ ] Implement `fft` and `ifft` using `nfft`. I like https://github.com/jakevdp/nfft
- [ ] `split`
- [ ] (some) ROIs:
    - [ ] `RectangularROI`
    - [ ] `SpanROI`
    - [ ] `Point2DROI`

The question is how to proceed. Should we releave v1.6 this week and include this in v1.7 or wait until the 4 main issues above (and any more that you may identify) are addressed to release v1.6?

ericpre commented:
Sounds like a very sensible plan to me.
Indeed, `DataAxis` needs to become binning aware and simply adding a `bin_width` attribute of the same shape as `axis` should be enough? But it also means that the generic functions using binning need to be adjusted accordingly.

francisco-dlp commented:
@jlaehne, for `FunctionalDataAxis` there is no need to store anything extra, but for `DataAxis` there is no way around it. Instead of creating a separate class for "binned" data axes, we could just add an `is_binned` attribute. Actually I think that we should deprecate `metadata.Signal.binned` since this is an attribute of the axes, not the signal. For example, if we transpose the signal and the navigation axes were not *binned* then the signal will no longer be "binned".

I agree with the argument of using the momentum of the release to get most of this done, but this inevitably means that we have to further delay v1.6 if you agree with the points of my previous post. We have to make sure that others agree.

ericpre commented:
I don't think I really answered the question... ;) Sort out binning signal for plotting, saving and fitting for v1.6 sounds sensible to me.

francisco-dlp commented:
OK, I'll point to this discussion in #2251 in case somebody argues against further delaying v1.6.

To proceed I suggest to

1. Merge this as-is in the `non-uniform-axes`
2. Open an issue to track the progress on the checklist above
3. Use the tracking issue to discuss how to best implement the features before starting work on it

francisco-dlp commented:
Tracking issue: #2398 

Do you agree to merge this into `non_uniform_axes` as-is?

ericpre commented:
Yes. Is there anything more you want to do in this PR?

jlaehne commented:
From my POV the issues that this PR was about were addressed, so please do merge with `non_uniform_axes`. And we can then launch new PRs. But it is important to have the new concept for FDA merged.

I will continue to contribute to the discussion in #2398.
